fili-core/src/main/java/com/yahoo/bard/webservice/data/filterbuilders/ConjunctionDruidFilterBuilder.java

```diff
@@ -88,6 +96,80 @@ public Filter buildFilters(Map<Dimension, Set<ApiFilter>> filterMap) throws Dime
     protected abstract Filter buildDimensionFilter(Dimension dimension, Set<ApiFilter> filters)
             throws DimensionRowNotFoundException;
 
+    /**
+     * Splits a set of {@link ApiFilter}s into two groups of positive and negative filters.
+     * <p>
+     * An {@link ApiFilter} is defined to be positive if its filter operation is one of
+     * <ul>
+     *     <li> {@link FilterOperation#in}
+     *     <li> {@link FilterOperation#startswith}
+     *     <li> {@link FilterOperation#contains}
+     *     <li> {@link FilterOperation#eq}
+     * </ul>
+     * An {@link ApiFilter} is defined to be negative if its filter operation is {@link FilterOperation#notin}.
```

Comment:
Can this split (positive/negative) be pushed to the type system to enforce?

Reply:
Sounds like a lot of specification overhead for minimal gain.

Converting between types requires *so much ceremony*.

Reply:
I don't disagree that Java makes type conversion painful some times, but I always like to let the compiler do the checking and thinking when possible via the type system instead of embedding what is essentially type information into system logic / rules.

I trust types and the compiler more than my own code ;-)

But, if it's not worth it and we think this logic is easy to understand and hard to screw up, then I agree that the gain from the compiler keeping it correct is not worth it.